wgsl/index.bs

```diff
@@ -865,135 +1031,47 @@ It is defined recursively in the following table:
           <!-- Vulkan base alignment: 4 times scalar aligment -->
   <tr algorithm="alignment of an array">
       <td>array<|T|,|N|>
-      <td>|Align|(|T|,`storage`)
-      <td>[=roundUp=](16, |Align|(|T|,`uniform`))
+      <td>[=RequiredAlignOf=](|T|, |C|)
+      <td>[=roundUp=](16, [=RequiredAlignOf=](|T|, |C|))
   <tr algorithm="alignment of an runtime-sized array">
       <td>array<|T|>
-      <td>|Align|(|T|,`storage`)
-      <td>[=roundUp=](16, |Align|(|T|,`uniform`))
+      <td>[=RequiredAlignOf=](|T|, |C|)
+      <td>[=roundUp=](16, [=RequiredAlignOf=](|T|, |C|))
   <tr algorithm="alignment of a structure">
-      <td>struct&lt;|T1|,...,|Tn|&gt;
-      <td>max(|Align|(|T1|,`storage`),..., |Align|(|Tn|,`storage`))
-      <td>[=roundUp=](16, |A|),<br> where |A| = max(|Align|(|T1|,`uniform`),..., |Align|(|Tn|,`uniform`)))
+      <td>struct&lt;T<sub>0</sub>, ... , T<sub>n</sub>&gt;
+      <td>max([=RequiredAlignOf=](T<sub>0</sub>, |C|), ... , [=RequiredAlignOf=](T<sub>n</sub>, |C|))
+      <td>[=roundUp=](16, |A|),<br>
+          where |A| = max([=RequiredAlignOf=](T<sub>0</sub>, |C|), ... , [=RequiredAlignOf=](T<sub>n</sub>, |C|)))
 </table>
 
-The <dfn>allocation extent</dfn> of a value |V|
-is the number of contiguous bytes reserved in host-shareable memory
-for the purpose of storing |V|.
-It is a function of the type of |V|, the size of any runtime-sized array that |V| may
-contain, and the [=storage class=] of the buffer.
+All structure members of type |T| must have a byte offset from the start of the
+structure that is a multiple of the [=RequiredAlignOf=](|T|, |C|) for the storage
+class |C|:
 
-Note: The allocation extent may include padding inserted to satisfy alignment rules.
-Consequently, loads and stores of a value might access fewer memory locations
-than value's allocation extent.
+<p algorithm="structure member minimum alignment">
+    [=OffsetOf=](|S|, |M|) = |k| &times; [=RequiredAlignOf=](|T|, C)<br>
+    Where |k| is a non-negative integer and |M| is a member of structure |S| with type |T|
+</p>
 
-We write |Extent|(|V|,|C|) for the allocation extent of value |V|
-of host-shareable type |S| in storage class |C|,
-where |C| is either [=storage classes/storage=] or [=storage classes/storage=].
-It is defined recursively in the following table:
+All arrays of element type |T| must have an element [=stride=] that is a
```

Comment:
Nit: should we describe maximum strides in terms of overall buffer sizes?

Reply:
Sorry, can you elaborate a bit?

Reply:
Somewhere we need a rule about stride that says the array must be able to fit in memory. I'm not sure where that rule belongs exactly, but this seemed as good a place as any. For example, if memory were 16 bytes total we could not have an array of 4 ints with a stride of 8.

Reply:
With the current state of the PR the size of the type (`struct`) cannot be explicitly set. The only way an array size may not fit in memory would be due to an array structure member being annotated with `[[size(n)]]` and the array member being larger than that.

We already have this statement in `Structure Layout Rules`:

> Structure members decorated with the [=size=] attribute must have a size at least as large as the size of the member's type. 

Which I think covers this case.